### PR TITLE
[FLINK-34572] Support OceanBase Jdbc Catalog

### PR DESCRIPTION
**Pull request by whhe**

### CODE REVIEW

**Comment by whhe** [comment]
Would you please take a look on this?  @leonardBang

**Comment by whhe** [comment]
@MartijnVisser PTAL

**Comment by whhe** [comment]
@eskabetxe PTAL

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseCatalog.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                        .isEmpty()
+                : !extractColumnValuesBySQL(
+                                getConnUrl(),
+                                "SELECT TABLE_NAME FROM ALL_TABLES "
+                                        + "WHERE OWNER = ? and TABLE_NAME = ?",
+                                1,
+                                null,
+                                tablePath.getDatabaseName(),
+                                tablePath.getObjectName())
+                        .isEmpty();
```
Comment (eskabetxe) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseCatalog.java:137]:
```
String query = isMySQLMode()
        ? "SELECT TABLE_NAME FROM information_schema.`TABLES` "
                        + "WHERE TABLE_SCHEMA = ? and TABLE_NAME = ?"
        : "SELECT TABLE_NAME FROM ALL_TABLES "
                        + "WHERE OWNER = ? and TABLE_NAME = ?";

return !extractColumnValuesBySQL(
                getConnUrl(),
                query,
                1,
                null,
                tablePath.getDatabaseName(),
                tablePath.getObjectName())
        .isEmpty();
```

File: flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseOracleCatalogITCase.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.flink.connector.jdbc.databases.oceanbase.OceanBaseOracleTestBase.tableRow;
+import static org.apache.flink.connector.jdbc.testutils.tables.TableBuilder.dbType;
+import static org.apache.flink.connector.jdbc.testutils.tables.TableBuilder.field;
+import static org.apache.flink.connector.jdbc.testutils.tables.TableBuilder.pkField;
+
+/** E2E test for {@link OceanBaseCatalog} with OceanBase Oracle mode. */
+@Disabled
```
Comment (eskabetxe) [inline flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseOracleCatalogITCase.java:41]: is this disable by some reason?
if yes, could we add some comment why..

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseCatalog.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                    getConnUrl(),
+                    "SELECT `SCHEMA_NAME` FROM `INFORMATION_SCHEMA`.`SCHEMATA`",
+                    1,
+                    dbName -> !builtinDatabases.contains(dbName));
+        }
+        return extractColumnValuesBySQL(
+                getConnUrl(),
+                "SELECT USERNAME FROM ALL_USERS",
+                1,
+                schema -> !builtinDatabases.contains(schema));
```
Comment (eskabetxe) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseCatalog.java:99]:
```
String query = isMySQLMode()
        ? "SELECT `SCHEMA_NAME` FROM `INFORMATION_SCHEMA`.`SCHEMATA`"
        : "SELECT USERNAME FROM ALL_USERS";

return extractColumnValuesBySQL(
        getConnUrl(),
        "SELECT USERNAME FROM ALL_USERS",
        1,
        dbName -> !builtinDatabases.contains(dbName));
```

**Review by eskabetxe — COMMENTED** [review]
I left some minor comments

**Comment by whhe** [comment]
> I left some minor comments

Hi @eskabetxe, I updated these files, please take a look again.

**Comment by whhe** [comment]
@eskabetxe PTAL

**Comment by whhe** [comment]
Anyone available? @eskabetxe @GOODBOY008 @leonardBang

**Review by RocMarshal — COMMENTED** [review]
Hi, @whhe Thx for the contribution.
Would you mind fixing the failed CI action before the next review ?

File: docs/content/docs/connectors/table/jdbc.md [hunk 4]
```diff
@@ -457,17 +456,19 @@ Other `Catalog` methods are currently not supported.
 
 ### Usage of JDBC Catalog
 
-The section mainly describes how to create and use a Postgres Catalog or MySQL Catalog.
+The section mainly describes how to create and use a Jdbc Catalog.
```
Comment (RocMarshal) [inline docs/content/docs/connectors/table/jdbc.md:459]:
```suggestion
The section mainly describes how to create and use a JDBC Catalog.
```

File: docs/content.zh/docs/connectors/table/jdbc.md [hunk 5]
```diff
@@ -433,10 +433,9 @@ JDBC Catalog
 
 `JdbcCatalog` 允许用户通过 JDBC 协议将 Flink 连接到关系数据库。
 
-目前，JDBC Catalog 有两个实现，即 Postgres Catalog 和 MySQL Catalog。目前支持如下 catalog 方法。其他方法目前尚不支持。
+目前，JDBC Catalog 有这几个实现：Postgres Catalog、MySQL Catalog、CrateDB Catalog 和 OceanBase Catalog。目前支持如下 catalog 方法。其他方法目前尚不支持。
```
Comment (RocMarshal) [inline docs/content.zh/docs/connectors/table/jdbc.md:436]:
```suggestion
目前，JDBC Catalog 有四个实现：Postgres Catalog、MySQL Catalog、CrateDB Catalog 和 OceanBase Catalog。目前支持如下 catalog 方法。其他方法目前尚不支持。
```

File: docs/content.zh/docs/connectors/table/jdbc.md [hunk 6]
```diff
@@ -451,17 +450,19 @@ tableExists(ObjectPath tablePath);
 
 ### JDBC Catalog 的使用
 
-本小节主要描述如果创建并使用 Postgres Catalog 或 MySQL Catalog。
+本小节主要描述如何创建并使用 Jdbc Catalog。
```
Comment (RocMarshal) [inline docs/content.zh/docs/connectors/table/jdbc.md:453]:
```suggestion
本小节主要描述如何创建并使用 JDBC Catalog。
```

**Review by RocMarshal — COMMENTED** [review]
Thanks @whhe for the contribution.
Left a few of comments. PTAL ~

**Comment by whhe** [comment]
> Thanks @whhe for the contribution. Left a few of comments. PTAL ~

Thanks for your review, the documentations have been updated as suggested. PTAL.

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseTypeMapper.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        String typeName = metadata.getColumnTypeName(colIndex).toUpperCase();
+        int jdbcType = metadata.getColumnType(colIndex);
+        String columnName = metadata.getColumnName(colIndex);
+        int precision = metadata.getPrecision(colIndex);
+        int scale = metadata.getScale(colIndex);
+        switch (jdbcType) {
+            case Types.BIT:
+                return DataTypes.BOOLEAN();
+            case Types.TINYINT:
+                if (isUnsignedType(typeName) || precision > 4) {
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseTypeMapper.java:60]: How about convert the lines like it to `return condition ? valueA : valueB;` ?

File: flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/databases/oceanbase/table/OceanBaseOracleDynamicTableSourceITCase.java [hunk 8]
```diff
@@ -32,11 +32,12 @@
 import java.util.Arrays;
 import java.util.List;
 
+import static org.apache.flink.connector.jdbc.databases.oceanbase.OceanBaseOracleTestBase.tableRow;
 import static org.apache.flink.connector.jdbc.testutils.tables.TableBuilder.dbType;
 import static org.apache.flink.connector.jdbc.testutils.tables.TableBuilder.field;
 
 /** The Table Source ITCase for OceanBase Oracle mode. */
-@Disabled
+@Disabled("OceanBase Oracle mode can only be tested locally.")
```
Comment (RocMarshal) [inline flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/databases/oceanbase/table/OceanBaseOracleDynamicTableSourceITCase.java:40]: 👍

**Review by RocMarshal — CHANGES_REQUESTED** [review]
Thanks @whhe for the update.
Left a few of comments.

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalogUtils.java [hunk 9]
```diff
@@ -82,6 +84,14 @@ public static AbstractJdbcCatalog createCatalog(
         } else if (dialect instanceof MySqlDialect) {
             return new MySqlCatalog(
                     userClassLoader, catalogName, defaultDatabase, baseUrl, connectionProperties);
+        } else if (dialect instanceof OceanBaseDialect) {
```
[the remark this answers is not in the document]
Comment (whhe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/JdbcCatalogUtils.java:87]: It seems better, will hava a try.

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseTypeMapper.java (hunk 7 above)
Comment (whhe) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/databases/oceanbase/catalog/OceanBaseTypeMapper.java:60]: OK

**Comment by whhe** [comment]
@RocMarshal Updated. PTAL

**Comment by RocMarshal** [comment]
@whhe nice work!
Could you help squish the commits into a single commit ? thx

**Review by RocMarshal — APPROVED** [review]
Thanks @whhe for the hard work!
LGTM +1, 
now ping @snuyanzin @eskabetxe

**Comment by whhe** [comment]
@leonardBang @MartijnVisser @GOODBOY008 Can someone help with the review?

**Comment by eskabetxe** [comment]
@whhe can you rebase please, and move the related code to the new OceanBase module.
we just merge a part of [FLIP-449](https://cwiki.apache.org/confluence/display/FLINK/FLIP-449%3A+Reorganization+of+flink-connector-jdbc)

If you need assistance, I'll be more than happy to help.

**Comment by whhe** [comment]
> @whhe can you rebase please, and move the related code to the new OceanBase module. we just merge a part of [FLIP-449](https://cwiki.apache.org/confluence/display/FLINK/FLIP-449%3A+Reorganization+of+flink-connector-jdbc)
> 
> If you need assistance, I'll be more than happy to help.

Thank you for your reminder. I will adjust the code as soon as possible and may trouble you to review it later.

**Comment by whhe** [comment]
@eskabetxe PTAL

File: flink-connector-jdbc-oceanbase/src/main/java/org/apache/flink/connector/jdbc/oceanbase/database/catalog/OceanBaseCatalog.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+            String defaultDatabase,
+            String baseUrl,
+            Properties connectionProperties) {
+        super(userClassLoader, catalogName, defaultDatabase, baseUrl, connectionProperties);
+        this.compatibleMode = compatibleMode;
+        this.dialectTypeMapper = new OceanBaseTypeMapper(compatibleMode);
+    }
+
+    private boolean isMySQLMode() {
+        return "mysql".equalsIgnoreCase(compatibleMode);
```
Comment (eskabetxe) [inline flink-connector-jdbc-oceanbase/src/main/java/org/apache/flink/connector/jdbc/oceanbase/database/catalog/OceanBaseCatalog.java:95]: should we have the compatible-mode on an enum?

File: flink-connector-jdbc-oceanbase/src/main/java/org/apache/flink/connector/jdbc/oceanbase/database/OceanBaseFactory.java [hunk 11]
```diff
@@ -54,6 +55,20 @@ public JdbcCatalog createCatalog(
             String username,
             String pwd,
             String baseUrl) {
-        throw new UnsupportedOperationException("Catalog for OceanBase is not supported yet.");
+        throw new UnsupportedOperationException(
+                "Catalog for OceanBase without 'compatible-mode' is not supported.");
```
[the remark this answers is not in the document]
Comment (whhe) [reply flink-connector-jdbc-oceanbase/src/main/java/org/apache/flink/connector/jdbc/oceanbase/database/OceanBaseFactory.java:58]: I rethought it and it seems that there should be a default value in createDialect and createCatalog. I will have a try.

**Review by 1996fanrui — COMMENTED** [review]
Thanks @whhe for the contribution, and thanks @RocMarshal & @eskabetxe for the review!

Merging

**Pull request merged**